Resolving the Thread-Safety Problem in registerReferenceKeyAndBeanName Using an Overloaded computeIfAbsent

juzzia opened the pull request:
org.apache.dubbo.config.spring.reference.ReferenceBeanManager.java
```java   
    public void registerReferenceKeyAndBeanName(String referenceKey, String referenceBeanNameOrAlias) {
        List<String> list =
                ConcurrentHashMapUtils.computeIfAbsent(referenceKeyMap, referenceKey, (key) -> new ArrayList<>());
        if (!list.contains(referenceBeanNameOrAlias)) {
            list.add(referenceBeanNameOrAlias);
            // register bean name as alias
            referenceAliasMap.put(referenceBeanNameOrAlias, list.get(0));
        }
    }
```
The ConcurrentHashMapUtils.computeIfAbsent method ensures the creation of a new list in a thread-safe manner when no list exists for the given referenceKey. However, the subsequent operations if (!list.contains(referenceBeanNameOrAlias)) and list.add(referenceBeanNameOrAlias) are not atomic. This situation can lead to thread-safety issues where multiple threads might simultaneously pass the contains check and then all perform the add operation, resulting in duplicate referenceBeanNameOrAlias entries within the list.

To resolve this issue, I added an overloaded version of the computeIfAbsent method that allows for thread-safe operations on the value through the use of a Consumer<T> functional interface. Additionally, I provided concurrent unit tests:

* ConcurrentHashMapUtilsTest#threadSafetyOperatorForJava8Test
* ConcurrentHashMapUtilsTest#threadSafetyOperatorForJava17Test
These tests confirm the correctness of the implementation.

The profile defined here cannot run the `spotless plugin` in JDK 8 because the version of the Spotless plugin is too high. Therefore, I've overridden the version of the Spotless plugin in the jdk8 - jdk11 profile to make it compatible with the lower - version environment.  `spotless plugin` version 2.22.0 can work properly in JDK versions ranging from 8 to 11.
```xml
    <profile>
      <id>jdk9-jdk11-spotless</id>
      <activation>
        <jdk>[1.8, 11)</jdk>
      </activation>
      <properties>
        <palantirJavaFormat.version>1.1.0</palantirJavaFormat.version>
      </properties>
    </profile>
```

```xml

    <properties>
      <spotless-maven-plugin.lower.version>2.22.0</spotless-maven-plugin.lower.version>
    </properties>
    <profile>
      <id>jdk8-jdk11-spotless</id>
      <activation>
        <jdk>[1.8, 11)</jdk>
      </activation>
      <properties>
        <palantirJavaFormat.version>1.1.0</palantirJavaFormat.version>
        <spotless-maven-plugin.version>${spotless-maven-plugin.lower.version}</spotless-maven-plugin.version>
      </properties>
    </profile>
```
Use a lower version of the spotless-maven-plugin dependency to override the version defined by spotless-maven-plugin.version so as to support JDK 1.8.

juzzia committed:
```java
    public void registerReferenceKeyAndBeanName(String referenceKey, String referenceBeanNameOrAlias) {
        List<String> list =
                ConcurrentHashMapUtils.computeIfAbsent(referenceKeyMap, referenceKey, (key) -> new ArrayList<>());
        if (!list.contains(referenceBeanNameOrAlias)) {
            list.add(referenceBeanNameOrAlias);
            // register bean name as alias
            referenceAliasMap.put(referenceBeanNameOrAlias, list.get(0));
        }
    }
```
The ConcurrentHashMapUtils.computeIfAbsent method ensures the creation of a new list in a thread-safe manner when no list exists for the given referenceKey. However, the subsequent operations if (!list.contains(referenceBeanNameOrAlias)) and list.add(referenceBeanNameOrAlias) are not atomic. This situation can lead to thread-safety issues where multiple threads might simultaneously pass the contains check and then all perform the add operation, resulting in duplicate referenceBeanNameOrAlias entries within the list.

To resolve this issue, I added an overloaded version of the computeIfAbsent method that allows for thread-safe operations on the value through the use of a Consumer<T> functional interface. Additionally, I provided concurrent unit tests:

ConcurrentHashMapUtilsTest#threadSafetyOperatorForJava8Test
ConcurrentHashMapUtilsTest#threadSafetyOperatorForJava17Test
These tests confirm the correctness of the implementation.

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.

juzzia committed:
Fix the errors caused by multiple executions of test methods in high - version JDK environments.

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.
😵

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.
😭

AlbumenJ commented:
![image](https://github.com/user-attachments/assets/ba16cac3-f675-427c-9318-9733cc6fe0c2)

juzzia commented:
> ![image](https://private-user-images.githubusercontent.com/9292748/421159957-ba16cac3-f675-427c-9318-9733cc6fe0c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3MDg2ODcsIm5iZiI6MTc0MTcwODM4NywicGF0aCI6Ii85MjkyNzQ4LzQyMTE1OTk1Ny1iYTE2Y2FjMy1mNjc1LTQyN2MtOTMxOC05NzMzY2M2ZmUwYzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMTU1MzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTJhMzM3MjM4YTM4Mzc2ZDlkNmM0ZTBlNjdjODM1ODdlMzYzMjUyMTJjYzYxNjM0OTZmOWFkYmRjYTdhNDg1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7suL5iheaJDXi_j3cGBqFwJNe4oEJeTY8S5lw--LiZE)

Thank you very much for the screenshot 👍. Due to continuous errors when downloading local Maven dependencies, the project still cannot run mvn spotless:apply. Therefore, I have been trying to push my code to check if my code formatting passes the checks 😭. I will use your screenshot as a reference and try to reformat my code again.

juzzia committed:
Format the code to comply with the project's code style guidelines and pass automated checks.

juzzia committed:
[fixbug]
 Solve the issue where the commands mvn spotless:apply and mvn spotless:check cannot be executed in a JDK 8 environment.

Use a lower version of the spotless-maven-plugin dependency to override the version defined by spotless-maven-plugin.version so as to support JDK 1.8.

juzzia committed:
Merge branch 'apache:3.3' into 3.3

juzzia commented:
@AlbumenJ I'm not sure why the workflow execution timed out since my code changes should not affect the test runs of the AbstractDynamicConfigurationTest class. I suspect it might be due to dependent services. Could you please rerun the PR action?

juzzia committed:
Using JavaDoc Tags to Optimize Comment Display

juzzia committed:
Merge branch '3.3' of https://github.com/juzzia/dubbo into 3.3

zrlw commented:
> @AlbumenJ I'm not sure why the workflow execution timed out since my code changes should not affect the test runs of the AbstractDynamicConfigurationTest class. I suspect it might be due to dependent services. Could you please rerun the PR action?

You can trigger it on your own Github Repo Actions like this,
![image](https://github.com/user-attachments/assets/a43c3258-b877-44a9-8f14-9259f67721c4)

zrlw commented:
since the author doesn't respond for a long time, i submit a new pr #15258
